Update asciidoctorj dependency

Current asciidoctorj dependency is of 1.5.4.1 version (it is already 3 years old).
See: https://mvnrepository.com/artifact/org.asciidoctor/asciidoctorj

I tried upgrading the AsciidoctorJ dependency to `2.1.0`. The `asciidoc-confluence-publisher-converter` module does compile, but then many of the unit tests fail.

<img width="1792" alt="Screen Shot 2019-11-02 at 15 50 58" src="https://user-images.githubusercontent.com/117866/68072689-97764180-fd88-11e9-8dcb-9c909078004a.png">

Here is one of the errors:

```
java.lang.RuntimeException: Could not convert asciidoc pages

	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluenceConverter.convert(AsciidocConfluenceConverter.java:91)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluenceConverterTest.convertAndBuildConfluencePages_withPageTitlePostProcessor_convertsTemplatesAndReturnsMetadata(AsciidocConfluenceConverterTest.java:102)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: org.asciidoctor.jruby.internal.AsciidoctorCoreException: org.jruby.exceptions.SyntaxError: (SyntaxError) /var/folders/10/z3hh1sx95hq4x3x1rk0ct5bw0000gn/T/junit6821148417943363469/junit5145785486266203985/templates/block_image.html.slim:1: syntax error, unexpected keyword_do
_buf = ''; _slim_controls1 = html_tag_if (attr? :link), :a, {href: (attr :link)} do; _slim_controls2 = ''; 

	at org.asciidoctor.jruby.internal.JRubyAsciidoctor.convert(JRubyAsciidoctor.java:314)
	at org.asciidoctor.jruby.internal.JRubyAsciidoctor.convert(JRubyAsciidoctor.java:330)
	at org.asciidoctor.jruby.internal.JRubyAsciidoctor.convert(JRubyAsciidoctor.java:325)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluencePage.convertedContent(AsciidocConfluencePage.java:129)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluencePage.newAsciidocConfluencePage(AsciidocConfluencePage.java:114)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluenceConverter.lambda$buildPageTree$0(AsciidocConfluenceConverter.java:103)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1080)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluenceConverter.buildPageTree(AsciidocConfluenceConverter.java:98)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluenceConverter.lambda$buildPageTree$0(AsciidocConfluenceConverter.java:109)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluenceConverter.buildPageTree(AsciidocConfluenceConverter.java:98)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluenceConverter.convert(AsciidocConfluenceConverter.java:82)
	... 33 more
Caused by: org.jruby.exceptions.SyntaxError: (SyntaxError) /var/folders/10/z3hh1sx95hq4x3x1rk0ct5bw0000gn/T/junit6821148417943363469/junit5145785486266203985/templates/block_image.html.slim:1: syntax error, unexpected keyword_do
_buf = ''; _slim_controls1 = html_tag_if (attr? :link), :a, {href: (attr :link)} do; _slim_controls2 = ''; 

	at org.jruby.RubyModule.module_eval(org/jruby/RubyModule.java:3255)
	at RUBY.compile_template_method(uri:classloader:/gems/tilt-2.0.9/lib/tilt/template.rb:274)
	at RUBY.compiled_method(uri:classloader:/gems/tilt-2.0.9/lib/tilt/template.rb:236)
	at org.jruby.ext.thread.Mutex.synchronize(org/jruby/ext/thread/Mutex.java:165)
	at RUBY.compiled_method(uri:classloader:/gems/tilt-2.0.9/lib/tilt/template.rb:235)
	at RUBY.evaluate(uri:classloader:/gems/tilt-2.0.9/lib/tilt/template.rb:169)
	at RUBY.render(uri:classloader:/gems/tilt-2.0.9/lib/tilt/template.rb:109)
	at RUBY.convert(uri:classloader:/gems/asciidoctor-2.0.10/lib/asciidoctor/converter/template.rb:109)
	at RUBY.convert(uri:classloader:/gems/asciidoctor-2.0.10/lib/asciidoctor/converter/composite.rb:28)
	at RUBY.convert(uri:classloader:/gems/asciidoctor-2.0.10/lib/asciidoctor/abstract_block.rb:75)
	at RUBY.content(uri:classloader:/gems/asciidoctor-2.0.10/lib/asciidoctor/abstract_block.rb:84)
	at org.jruby.RubyArray.map(org/jruby/RubyArray.java:2577)
	at RUBY.content(uri:classloader:/gems/asciidoctor-2.0.10/lib/asciidoctor/abstract_block.rb:84)
	at RUBY.content(uri:classloader:/gems/asciidoctor-2.0.10/lib/asciidoctor/document.rb:1020)
	at RUBY.singleton class(/var/folders/10/z3hh1sx95hq4x3x1rk0ct5bw0000gn/T/junit6821148417943363469/junit5145785486266203985/templates/embedded.html.slim:7)
	at org.jruby.RubyBasicObject.instance_eval(org/jruby/RubyBasicObject.java:2609)
	at RUBY.singleton class(/var/folders/10/z3hh1sx95hq4x3x1rk0ct5bw0000gn/T/junit6821148417943363469/junit5145785486266203985/templates/embedded.html.slim:1)
	at RUBY.__tilt_2000(/var/folders/10/z3hh1sx95hq4x3x1rk0ct5bw0000gn/T/junit6821148417943363469/junit5145785486266203985/templates/embedded.html.slim:1)
	at org.jruby.RubyMethod.call(org/jruby/RubyMethod.java:120)
	at RUBY.evaluate(uri:classloader:/gems/tilt-2.0.9/lib/tilt/template.rb:170)
	at RUBY.render(uri:classloader:/gems/tilt-2.0.9/lib/tilt/template.rb:109)
	at RUBY.convert(uri:classloader:/gems/asciidoctor-2.0.10/lib/asciidoctor/converter/template.rb:109)
	at RUBY.convert(uri:classloader:/gems/asciidoctor-2.0.10/lib/asciidoctor/converter/composite.rb:28)
	at RUBY.convert(uri:classloader:/gems/asciidoctor-2.0.10/lib/asciidoctor/document.rb:951)
	at RUBY.convert(uri:classloader:/gems/asciidoctor-2.0.10/lib/asciidoctor/convert.rb:120)
```

Core problem appears to be summarized in this message:
```
syntax error, unexpected keyword_do
_buf = ''; _slim_controls1 = html_tag_if (attr? :link), :a, {href: (attr :link)} do; _slim_controls2 = ''; 
```

…and then there is a reference to this file:
- `block_image.html.slim`

However, when I check that file, all it contains is this:
```
= html_tag_if (attr? :link), :a, {href: (attr :link)}
  ac:image ac:height=(attr :height) ac:width=(attr :width) ac:title=(title if title?) ac:alt=(title if title?)
    - if uriish? attr :target
      ri:url ri:value=(attr :target)
    - else
      ri:attachment ri:filename=(attr :target)
  - if title?
    div class="cp-image-title"
      em =captioned_title
```

I don't see any `keyword_do` or `do` statement.

I wonder why these Slim templates now fail. Perhaps the Slim version was updated?

I'm diving into the first regression. Here's my analysis results.

The failure is in the class `AsciidocConfluencePageTest`, in the method named `renderConfluencePage_asciiDocWithInternalCrossReferenceInlineToSectionWithInlineAnchor_returnsConfluencePageContentWithInternalCrossReferenceToSectionUsingSectionTitle`:
```
    @Test
    public void renderConfluencePage_asciiDocWithInternalCrossReferenceInlineToSectionWithInlineAnchor_returnsConfluencePageContentWithInternalCrossReferenceToSectionUsingSectionTitle() {
        // arrange
        String adocContent = "" +
                "== Section 1 [[section-1]]\n" +
                "Cross reference to <<section-1>>";

        // act
        AsciidocConfluencePage asciidocConfluencePage = newAsciidocConfluencePage(asciidocPage(prependTitle(adocContent)), UTF_8, TEMPLATES_FOLDER, dummyAssetsTargetPath());

        // assert
        String expectedContent = "" +
                "<h1><ac:structured-macro ac:name=\"anchor\"><ac:parameter ac:name=\"\">section-1</ac:parameter></ac:structured-macro>Section 1</h1>" +
                "<p>Cross reference to <ac:link ac:anchor=\"section-1\"><ac:plain-text-link-body><![CDATA[Section 1]]></ac:plain-text-link-body></ac:link></p>";
        assertThat(asciidocConfluencePage.content(), is(expectedContent));
    }
```

The failure is on the last statement (the `assertThat`):
```
java.lang.AssertionError: 
Expected: is "<h1><ac:structured-macro ac:name=\"anchor\"><ac:parameter ac:name=\"\">section-1</ac:parameter></ac:structured-macro>Section 1</h1><p>Cross reference to <ac:link ac:anchor=\"section-1\"><ac:plain-text-link-body><![CDATA[Section 1]]></ac:plain-text-link-body></ac:link></p>"
     but: was "<h1><ac:structured-macro ac:name=\"anchor\"><ac:parameter ac:name=\"\">section-1</ac:parameter></ac:structured-macro>Section 1</h1><p>Cross reference to <ac:link ac:anchor=\"section-1\"><ac:plain-text-link-body><![CDATA[[section-1]]]></ac:plain-text-link-body></ac:link></p>"
Expected :<h1><ac:structured-macro ac:name=\"anchor\"><ac:parameter ac:name=\"\">section-1</ac:parameter></ac:structured-macro>Section 1</h1><p>Cross reference to <ac:link ac:anchor=\"section-1\"><ac:plain-text-link-body><![CDATA[Section 1]]></ac:plain-text-link-body></ac:link></p>
Actual   :<h1><ac:structured-macro ac:name=\"anchor\"><ac:parameter ac:name=\"\">section-1</ac:parameter></ac:structured-macro>Section 1</h1><p>Cross reference to <ac:link ac:anchor=\"section-1\"><ac:plain-text-link-body><![CDATA[[section-1]]]></ac:plain-text-link-body></ac:link></p>

	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluencePageTest.renderConfluencePage_asciiDocWithInternalCrossReferenceInlineToSectionWithInlineAnchor_returnsConfluencePageContentWithInternalCrossReferenceToSectionUsingSectionTitle(AsciidocConfluencePageTest.java:1177)
	[...]
```

The difference comes down to the link text content:
- Expected: `Section 1`
- Actual - `[section-1]`

Tracing the origins, I find this in the `inline_anchor.html.slim` file:
```
- case type
- when :xref
  - if cross_page_xref? target
    ac:link
      ri:page ri:content-title=target
      ac:plain-text-link-body
        <![CDATA[#{(xref_text || %([#{attr :refid}]))}]]>
  - else
    ac:link ac:anchor=(anchor_name @target)
      ac:plain-text-link-body
        <![CDATA[#{(xref_text || %([#{attr :refid}]))}]]>
```

The very last statement is what changed results between the Asciidoctor 1.5 and 2.1 versions.

It seems (but here some guess work is involved) that `xref_text` previously returned a non-`nil` value, but now it returns `nil`, causing the fallback `%([#{attr :refid}])` to kick in.

One level deeper, we find the `xref_text` function defined in `helpers.rb` as follows:
```
  def xref_text
    str = text || document.references[:ids][attr :refid || target]
    str.tr_s("\n", ' ') if str
  end
```

Apparently, `document.references` now no longer contains the anchor? (Not that `document.references` is defined as an alias for `document.catalog`).

@cstettler : Does this make sense so far?

--- a/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/_toc.html.slim
+++ b/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/_toc.html.slim
@@ -1,3 +1,3 @@
 ac:structured-macro ac:name="toc"
   ac:parameter ac:name="maxLevel"
-    =((attr? :toclevels) ? (attr :toclevels) : 2)
+    = attr :toclevels, 2, true

--- a/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/block_image.html.slim
+++ b/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/block_image.html.slim
@@ -1,4 +1,4 @@
-= html_tag_if (attr? :link), :a, {href: (attr :link)}
+= html_a_tag_if (attr? :link)
   ac:image ac:height=(attr :height) ac:width=(attr :width) ac:title=(title if title?) ac:alt=(title if title?)
     - if uriish? attr :target
       ri:url ri:value=(attr :target)

--- a/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/helpers.rb
+++ b/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/helpers.rb
@@ -1,12 +1,12 @@
 require 'asciidoctor'
 require 'json'
 
-if Gem::Version.new(Asciidoctor::VERSION) <= Gem::Version.new('1.5.1')
-  fail 'asciidoctor: FAILED: HTML5/Slim backend needs Asciidoctor >=1.5.2!'
+if Gem::Version.new(Asciidoctor::VERSION) <= Gem::Version.new('2.0.0')
+  fail 'asciidoctor: FAILED: HTML5/Slim backend needs Asciidoctor >=1.0.0!'
 end
 
 unless defined? Slim::Include
-  fail 'asciidoctor: FAILED: HTML5/Slim backend needs Slim >= 2.1.0!'
+  fail 'asciidoctor: FAILED: HTML5/Slim backend needs Slim >= 4.0.0!'
 end
 
 # Add custom functions to this module that you want to use in your Slim
@@ -64,13 +64,12 @@ module Slim::Helpers
   end
 
   ##
-  # Conditionally wraps a block in an element. If condition is +true+ then it
-  # renders the specified tag with optional attributes and the given
-  # block inside, otherwise it just renders the block.
+  # Conditionally wraps a block in an a element. If condition is +true+ then it
+  # renders the a tag and the given block inside, otherwise it just renders the block.
   #
   # For example:
   #
-  #    = html_tag_if link?, 'a', {class: 'image', href: (attr :link)}
+  #    = html_a_tag_if link?
   #      img src='./img/tux.png'
   #
   # will produce:
@@ -86,15 +85,13 @@ module Slim::Helpers
   # otherwise.
   #
   # @param condition [Boolean] the condition to test to determine whether to
-  #        render the enclosing tag.
-  # @param name (see #html_tag)
-  # @param attributes (see #html_tag)
+  #        render the enclosing a tag.
   # @yield (see #html_tag)
   # @return [String] a rendered HTML fragment.
   #
-  def html_tag_if(condition, name, attributes = {}, &block)
+  def html_a_tag_if(condition, &block)
     if condition
-      html_tag name, attributes, &block
+      html_tag :a, {href: (attr :link)}, &block
     else
       yield
     end
@@ -148,7 +145,12 @@ module Slim::Helpers
   #
   # @return [String, nil] text of the xref anchor, or +nil+ if not found.
   def xref_text
-    str = text || document.references[:ids][attr :refid || target]
+    if attributes[:refid] == text
+      ref = document.catalog[:refs][attributes['refid'] || target]
+      str = (ref ? ref.xreftext : text)
+    else
+      str = text
+    end
     str.tr_s("\n", ' ') if str
   end
 

--- a/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/inline_anchor.html.slim
+++ b/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/inline_anchor.html.slim
@@ -6,13 +6,13 @@
       ac:plain-text-link-body
         <![CDATA[#{(xref_text || %([#{attr :refid}]))}]]>
   - else
-    ac:link ac:anchor=(anchor_name @target)
+    ac:link ac:anchor=(anchor_name target || id)
       ac:plain-text-link-body
         <![CDATA[#{(xref_text || %([#{attr :refid}]))}]]>
 - when :ref
   ac:structured-macro ac:name="anchor"
     ac:parameter ac:name=""
-     =(anchor_name target)
+     =(anchor_name target || id)
 - when :link
   - if uriish? target
     a href=target

--- a/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/inline_image.html.slim
+++ b/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/inline_image.html.slim
@@ -1,4 +1,4 @@
-= html_tag_if (attr? :link), :a, {href: (attr :link)}
+= html_a_tag_if (attr? :link)
   ac:image ac:height=(attr :height) ac:width=(attr :width) ac:alt=(attr :alt if attr :alt)
     - if uriish? target
       ri:url ri:value=(target)

--- a/pom.xml
+++ b/pom.xml
@@ -165,13 +165,13 @@
             <dependency>
                 <groupId>org.asciidoctor</groupId>
                 <artifactId>asciidoctorj</artifactId>
-                <version>1.5.8.1</version>
+                <version>2.4.0</version>
                 <scope>compile</scope>
             </dependency>
             <dependency>
                 <groupId>org.asciidoctor</groupId>
                 <artifactId>asciidoctorj-diagram</artifactId>
-                <version>1.5.18</version>
+                <version>2.0.2</version>
                 <scope>compile</scope>
             </dependency>
             <dependency>

